File: 1387-find-elements-in-a-contaminated-binary-tree/find-elements-in-a-contaminated-binary-tree.py


class FindElements:
    def __init__(self, root: TreeNode):
        self.root = root

    def getPath(self, target: int) -> str:
        path=[]
        while target>0:
            path.append("L" if target%2 else "R")
            target=(target-1)//2 if target%2 else (target-2)//2
        return reversed(path)

        # reverse_array=[]
        # for i in range(-1,(-1*len(path))-1,-1):
        #     reverse_array.append(path[i])
        # return reverse_array
        # return "".join(reversed(path))
        # return "".join(reverse_array))
      

    def find(self, target: int) -> bool:
        path=self.getPath(target)
        node=self.root
        for i in path:
            if i=="L" and node.left:
                node=node.left
            elif i=="R" and node.right:
                node=node.right
            else:
                return False

        return True
       
# Usage:
# obj = FindElements(root)
# param_1 = obj.find(target)

File: 1387-find-elements-in-a-contaminated-binary-tree/find-elements-in-a-contaminated-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
//  public class TreeNode {
//       int val;
//       TreeNode left;
//      TreeNode right;
//       TreeNode() {}
//       TreeNode(int val) { this.val = val; }
//       TreeNode(int val, TreeNode left, TreeNode right) {
//           this.val = val;
//           this.left = left;
//           this.right = right;
//       }
//   }

import java.util.*;
class FindElements {
    HashSet<Integer> seen=new HashSet<>();
    TreeNode root;
    int val=0;
    private void dfs(TreeNode node,int val){
            seen.add(val);
            if (node.left!=null){
                dfs(node.left,(val*2)+1);
            }
            if (node.right!=null){
                dfs(node.right,(val*2)+2);
            }

        }
    public FindElements(TreeNode root) {
        this.root=root;
        dfs(root,val);
        System.out.println(seen);

        
    }
        

    
    public boolean find(int target) {
        if(seen.contains(target)){
            return true;
        }
        else{
            return false;
        }

        
    }
}

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements obj = new FindElements(root);
 * boolean param_1 = obj.find(target);
 */